plugins/LanguagesManager/Commands/Update.php

```diff
@@ -55,79 +48,78 @@ protected function execute(InputInterface $input, OutputInterface $output)
-            }
+                $code = basename($filename, '.json');
+
+                if (!in_array($code, $languageCodes)) {
+
+                    if (!empty($plugin)) {
+                        continue; # never create a new language for plugin only
+                    }
+
+                    $createNewFile = false;
+                    if ($input->isInteractive()) {
+                        $createNewFile = $dialog->askConfirmation($output, "\nLanguage $code does not exist. Should it be added? ", false);
+                    }
+
+                    if (!$createNewFile) {
+                        continue; # do not create a new file for the language
+                    }
 
-            $command = $this->getApplication()->find('translations:set');
-            $arguments = array(
-                'command'  => 'translations:set',
-                '--code'   => $code,
-                '--file'   => $filename,
-                '--plugin' => $plugin
-            );
-            $inputObject = new ArrayInput($arguments);
-            $inputObject->setInteractive($input->isInteractive());
-            $command->run($inputObject, new NullOutput());
-
-            // update core modules that aren't in their own repo
-            if (empty($plugin)) {
-
-                foreach (self::getPluginsInCore() as $pluginName) {
-
-                    // update translation files
-                    $command = $this->getApplication()->find('translations:set');
-                    $arguments = array(
-                        'command'  => 'translations:set',
-                        '--code'   => $code,
-                        '--file'   => $filename,
-                        '--plugin' => $pluginName
-                    );
-                    $inputObject = new ArrayInput($arguments);
-                    $inputObject->setInteractive($input->isInteractive());
-                    $command->run($inputObject, new NullOutput());
+                    @touch(PIWIK_DOCUMENT_ROOT . DIRECTORY_SEPARATOR . 'lang' . DIRECTORY_SEPARATOR . $code . '.json');
+                    API::unsetInstance(); // unset language manager instance, so valid names are refetched
                 }
+
+                $command = $this->getApplication()->find('translations:set');
+                $arguments = array(
+                    'command' => 'translations:set',
+                    '--code' => $code,
+                    '--file' => $filename,
+                    '--plugin' => $plugin
+                );
+                $inputObject = new ArrayInput($arguments);
+                $inputObject->setInteractive($input->isInteractive());
+                $command->run($inputObject, new NullOutput());
             }
+
+            $progress->finish();
         }
 
-        $progress->finish();
-        $output->writeln("Finished.");
+        $output->writeln("Finished in " . round(microtime(true)-$start, 3) . "s");
```

Comment:
FYI there is `Piwik\Timer` that can do that (don't need to change that code, it's just FYI in case you didn't know about it)